content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,172 @@
+This can hinder understanding of the issue.
+
+Because of this, `stopifnot()` was improved in R 4.0.0:
+
+> stopifnot() now allows customizing error messages via argument names, thanks to a patch proposal by Neal Fultz in PR\#17688.
+
+This means we can now provide a clearer error message directly in `stopifnot()` [^1]:
+
+[^1]: Read [the tidyverse style guide](https://style.tidyverse.org/error-messages.html) for more guidance on how to write good error messages.
+
+```{r, error = TRUE}
+say_hello <- function(name) {
+  stopifnot("`name` must be a character." = is.character(name))
+  paste("Hello", name)
+}
+
+say_hello(404)
+```
+
+This is clearly a really great improvement to the functionality of base R. However, we can see from this example that we could create the error message programmatically based on the contents of the test.
+Each time we test if the object is of `class_X` and this is not true, we could throw an error saying something like "x must of a class_X".
+This way, you don't have to repeat yourself which is generally a good aim [^2].
+This becomes necessary when you start having many input checks in your function or in your package.
+
+[^2]: The [Don't Repeat Yourself (DRY) principle of software development](https://en.wikipedia.org/wiki/Don't_repeat_yourself), also mentioned in this post on [caching](https://blog.r-hub.io/2021/07/30/cache/)
+
+## Checking function inputs using R packages
+
+### The example of the checkmate package
+
+Although some developers create [their own functions](https://github.com/djnavarro/bs4cards/blob/a021d731a307ec7af692a42364308b60e2bf9827/R/validators.R) to solve this problem, you can also rely on existing packages to make your life easier.
+One of these packages designed to help you in input checking is [checkmate](https://mllg.github.io/checkmate/).
+checkmate provides a large number of function to check that inputs respect a given set of properties, and returns clear error messages when that is not the case:
+
+```{r}
+say_hello <- function(name) {
+  # Among other things, check_string() checks that we provide a 
+  # character object of length one
+  checkmate::assert_string(name)
+  paste("Hello", name)
+}
+```
+
+```{r, error = TRUE}
+say_hello(404)
+```
+
+```{r, error = TRUE}
+say_hello(c("Bob", "Alice"))
+```
+
+### Other packages to check function inputs
+
+Because input checking is such an important point task and because it is so difficult to get right, it is not surprising that there are many packages other than checkmate to solve this issue.
+We will not get into the details of all of the available options here but below is a list of some of the them. If interested in understanding the various approaches to input taking the documentation
+for these package is a great place to start.
+
+-   testthat
+-   assertthat
+-   check
+-   assertr
+-   assertive
+-   ensurer
+-   `vctrs::vec_assert()`
+
```

Comment:
In the post I'd be curious to also hear about your motivation to write this and a short paragraph or so about what you ended up using in your package(s) and whether you got some user feedback.

Reply:
In short: some of use were tired of filling our functions with `stopifnot()` / `if () { stop() }` or writing the same kind of checking function over and over and we started collectively looking for existing alternatives.

In which part of the post do you think this could fit?

Reply:
Maybe right before the introduction? Something like

> Are you, like we were, tired of filling your functions with blabla?